source/extensions/filters/common/ext_authz/ext_authz.h

```diff
@@ -73,6 +73,8 @@ enum class CheckStatus {
   Denied
 };
 
+using UnvalidatedHeaderVector = std::vector<std::pair<std::string, std::string>>;
```

Comment:
Thanks for working on this!

Just a high level comment first:
It looks to me you are copying the response header as it is into ext_authz data structure, then trying to validate it when ApplyHeaderMutation().  I am wondering can you do the validation before copying the response header into ext_authz data structure. That way, you don't need to modify these data types in ext_authz.h.

Reply:
You are correct, that is what I'm doing. I considered doing it in a different way that didn't require changing this header, but I don't like the idea of adding the same validation code to both clients.

Reply:
@ggreenway @tyxia do you have some comments on this change?

Reply:
Understood there might be a little bit code duplication.  However, if you do it early on, i.e, in the client, you are able to avoid unnecessary header copy and storage, right?

Reply:
I think this only causes more copies for the `response_headers_to_...` fields, since instead of moving the entire vector, it only copies the valid headers. For all other fields, there are the same amount of copies: one copy into the "internal" Response struct, then again into the HeaderMap. 

I think there may be room to optimize the copies, for example by cannibalizing the response proto in the grpc client's onSuccess.

@tyxia @ggreenway @esmet Thoughts?

Reply:
Hm. On second thought, the validation might only be necessary in the grpc client anyway. In the raw http client, the headers are being sent over http, so added headers should always be valid.

One exception is that the side stream could add an invalid header to remove (since it's sent as a header VALUE), but that is easy enough to validate.

Reply:
That's a good point.  For HTTP response received by ext_authz HTTP client from the side stream, whether we should validate is a good question.  I assume Envoy HTTP codec/parser already validated them.  Please confirm this.

Reply:
This had the unintended side effect of adding validation to the network/ext_authz filter so I'm going to move the validation code back to the filter.

The network/ext_authz filter doesn't read or care about header / query parameter mutations, so it makes no sense for the client to ever reject the response because them. The mutation validation is only needed in the http filter, so the code should go in the http filter.